src/Features/CSharp/Portable/CodeFixes/AddExplicitCast/AddExplicitCastCodeFixProvider.cs

```diff
@@ -0,0 +1,372 @@
+        {
+            var castExpression = targetNode.Cast(conversionType);
+            var newRoot = currentRoot.ReplaceNode(targetNode, castExpression.WithAdditionalAnnotations(Simplifier.Annotation));
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        /// <summary>
+        /// Output the current type info of the target node and the conversion type(s) that the target node is going to be cast by.
+        /// Implicit downcast can appear on Variable Declaration, Return Statement, and Function Invocation, for example:
+        /// Base b; Derived d = [||]b;       
+        /// object b is the current node with type *Base*, and the conversion type which object b is going to be cast by is *Derived*
+        /// </summary>
+        /// <param name="semanticModel"></param>
+        /// <param name="root">The root of the tree of nodes.</param>
+        /// <param name="targetNode">The node to be cast.</param>
+        /// <param name="cancellationToken"></param>
+        /// <param name="targetNodeType">Output the type of <paramref name="targetNode"/>.</param>
+        /// <param name="potentialConversionTypes">>Output the potential conversions types that <paramref name="targetNode"/> can be cast to</param>
+        /// <returns>
+        /// True, if the target node has at least one conversion type, and they are assigned to <paramref name="potentialConversionTypes"/>
+        /// False, if the target node has no conversion type.
+        /// </returns>
+        private static bool GetTypeInfo(SemanticModel semanticModel, SyntaxNode root, SyntaxNode? targetNode, CancellationToken cancellationToken,
+            out ITypeSymbol? targetNodeType, out ImmutableArray<ITypeSymbol> potentialConversionTypes)
+        {
+            targetNodeType = null;
+            potentialConversionTypes = ImmutableArray<ITypeSymbol>.Empty;
+            var mutablePotentialConversionTypes = new List<ITypeSymbol>();
+            if (targetNode == null)
+            {
+                return false;
+            }
+
+            var targetNodeInfo = semanticModel.GetTypeInfo(targetNode, cancellationToken);
+            targetNodeType = targetNodeInfo.Type;
+
+            if (targetNodeType == null)
+            {
+                return false;
+            }
+
+            if (targetNodeInfo.ConvertedType != null && !targetNodeType.Equals(targetNodeInfo.ConvertedType))
+            {
+                mutablePotentialConversionTypes.Add(targetNodeInfo.ConvertedType);
+            }
+
+            if (targetNode.GetAncestors<ArgumentSyntax>().FirstOrDefault() is ArgumentSyntax targetArgument &&
+                targetArgument.Parent is ArgumentListSyntax argumentList && argumentList.Parent is SyntaxNode invocationNode) // invocation node could be Invocation Expression, Object Creation, Base Constructor...
```

Comment:
`targetNode.GetAncestors<ArgumentSyntax>().FirstOrDefault()` doesn't seem appropriate.  It could walk up far too high in the tree.  for example, say i have this:

```
Foo(() => {
     string InnerFunc(obj o) { return o; }
});
```

There will be an error returning `o` in the inner function.  But this check will find that we're in a lambda passed to `Foo`.

Reply:
> targetNode.GetAncestors<ArgumentSyntax>().FirstOrDefault() doesn't seem appropriate.  It could walk up far too high in the tree.  for example, say i have this:
> Foo(() => {
>      string InnerFunc(obj o) { return o; }
> });
> 
> There will be an error returning o in the inner function.  But this check will find that we're in a lambda passed to Foo.

Ah, this case is pretty helpful, I fix it in my latest code by checking whether the error happens on an assignment operation or happens on an invocation expression.